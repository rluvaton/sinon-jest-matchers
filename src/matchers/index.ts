export { sinonToBeCalled } from './sinonToBeCalled';
export { sinonToBeCalledTimes } from './sinonToBeCalledTimes';
export { sinonToBeCalledWith } from './sinonToBeCalledWith';
export { sinonLastCalledWith } from './sinonLastCalledWith';
export { sinonLastReturnedWith } from './sinonLastReturnedWith';
export { sinonNthCalledWith } from './sinonNthCalledWith';
export { sinonToReturn } from './sinonToReturn';
export { sinonToReturnTimes } from './sinonToReturnTimes';
export { sinonToReturnWith } from './sinonToReturnWith';
