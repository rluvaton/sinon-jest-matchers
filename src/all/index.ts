import * as matchers from '../matchers';

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
const jestExpect = global.expect;

if (jestExpect !== undefined) {
  jestExpect.extend(matchers);
} else {
  throw new Error(
    "Unable to find Jest's global expect. " +
      'Please check you have added sinon-jest-matchers correctly to your jest configuration. ' +
      'See https://github.com/rluvaton/sinon-jest-matchers#setup for help.',
  );
}
