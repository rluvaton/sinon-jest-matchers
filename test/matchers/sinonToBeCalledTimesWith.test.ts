import sinon from 'sinon';
import * as matchers from '../../src';
import * as Immutable from 'immutable';

expect.extend(matchers);

expect.extend({
  optionalFn(fn?: unknown) {
    const pass = fn === undefined || typeof fn === 'function';
    return { message: () => 'expect either a function or undefined', pass };
  },
});

describe('sinonToBeCalledTimesWith', () => {
  test('works only on spies or sinon.spy', () => {
    const fn = function fn() {};

    expect(() => expect(fn).sinonToBeCalledTimesWith(0)).toThrowErrorMatchingSnapshot();
  });

  test('works when not called', () => {
    const fn = sinon.spy();
    expect(fn).not.sinonToBeCalledTimesWith(2, 'foo', 'bar');

    expect(() => expect(fn).sinonToBeCalledTimesWith(2, 'foo', 'bar')).toThrowErrorMatchingSnapshot();
  });

  test('works with only number of called argument', () => {
    const fn = sinon.spy();
    fn();
    fn();
    expect(fn).sinonToBeCalledTimesWith(2);
  });

  test("works with arguments that don't match", () => {
    const fn = sinon.spy();
    fn('foo', 'bar');
    fn('foo', 'bar1');

    expect(fn).not.sinonToBeCalledTimesWith(2, 'foo', 'bar');

    expect(() => expect(fn).sinonToBeCalledTimesWith(2, 'foo', 'bar')).toThrowErrorMatchingSnapshot();
  });

  test("works with arguments that don't match in number of arguments", () => {
    const fn = sinon.spy();
    fn('foo', 'bar');
    fn('foo', 'bar', 'plop');
    fn('foo', 'bar');

    expect(fn).not.sinonToBeCalledTimesWith(3, 'foo', 'bar');

    expect(() => expect(fn).sinonToBeCalledTimesWith(3, 'foo', 'bar')).toThrowErrorMatchingSnapshot();
  });

  test("works with arguments that don't match with matchers", () => {
    const fn = sinon.spy();
    fn('foo', 1);
    fn('foo', 'bar');

    expect(fn).not.sinonToBeCalledTimesWith(2, expect.any(String), expect.any(Number));

    expect(() => expect(fn).sinonToBeCalledTimesWith(2, expect.any(String), expect.any(Number))).toThrowErrorMatchingSnapshot();
  });

  test("works with arguments that don't match with matchers even when argument is undefined", () => {
    const fn = sinon.spy();
    fn('foo', undefined);

    expect(fn).not.sinonToBeCalledTimesWith(1, 'foo', expect.any(String));

    expect(() => expect(fn).sinonToBeCalledTimesWith(1, 'foo', expect.any(String))).toThrowErrorMatchingSnapshot();
  });

  // jest issue 12463
  test.skip("works with arguments that don't match in size even if one is an optional matcher", () => {
    const fn = sinon.spy();
    fn('foo');

    expect(fn).not.sinonToBeCalledTimesWith(1, 'foo', (expect as any).optionalFn());
    expect(() => expect(fn).sinonToBeCalledTimesWith(1, 'foo', (expect as any).optionalFn())).toThrowErrorMatchingSnapshot();
  });

  test('works with arguments that match', () => {
    const fn = sinon.spy();
    fn('foo', 'bar');
    fn('foo', 'bar');

    expect(fn).sinonToBeCalledTimesWith(2, 'foo', 'bar');

    expect(() => expect(fn).not.sinonToBeCalledTimesWith(2, 'foo', 'bar')).toThrowErrorMatchingSnapshot();
  });

  test('works with arguments that match with matchers', () => {
    const fn = sinon.spy();
    fn('foo', 'bar');
    fn('foo', 'bar');

    expect(fn).sinonToBeCalledTimesWith(2, expect.any(String), expect.any(String));

    expect(() =>
      expect(fn).not.sinonToBeCalledTimesWith(2, expect.any(String), expect.any(String)),
    ).toThrowErrorMatchingSnapshot();
  });

  // Test fails, will ignore for now until it will because an actual problem
  test.skip('works with trailing undefined arguments', () => {
    const fn = sinon.spy();
    fn('foo', undefined);
    fn('foo', undefined);

    expect(() => expect(fn).sinonToBeCalledTimesWith(2, 'foo')).toThrowErrorMatchingSnapshot();
  });

  test('works with trailing undefined arguments if requested by the match query', () => {
    const fn = sinon.spy();
    fn('foo', undefined);
    fn('foo', undefined);

    expect(fn).sinonToBeCalledTimesWith(2, 'foo', undefined);
    expect(() => expect(fn).not.sinonToBeCalledTimesWith(2, 'foo', undefined)).toThrowErrorMatchingSnapshot();
  });

  test('works with trailing undefined arguments when explicitly requested as optional by matcher', () => {
    const fn = sinon.spy();
    fn('foo', undefined);
    fn('foo', undefined);

    expect(fn).sinonToBeCalledTimesWith(2, 'foo', (expect as any).optionalFn());
    expect(() =>
      expect(fn).not.sinonToBeCalledTimesWith(2, 'foo', (expect as any).optionalFn()),
    ).toThrowErrorMatchingSnapshot();
  });

  test('works with Map', () => {
    const fn = sinon.spy();

    const m1 = new Map([
      [1, 2],
      [2, 1],
    ]);
    const m2 = new Map([
      [1, 2],
      [2, 1],
    ]);
    const m3 = new Map([
      ['a', 'b'],
      ['b', 'a'],
    ]);

    fn(m1);
    fn(m1);
    fn(m3);

    expect(fn).sinonToBeCalledTimesWith(2, m2);
    expect(fn).not.sinonToBeCalledTimesWith(2, m3);

    expect(() => expect(fn).not.sinonToBeCalledTimesWith(2, m2)).toThrowErrorMatchingSnapshot();
    expect(() => expect(fn).sinonToBeCalledTimesWith(2, m3)).toThrowErrorMatchingSnapshot();
  });

  test('works with Set', () => {
    const fn = sinon.spy();

    const s1 = new Set([1, 2]);
    const s2 = new Set([1, 2]);
    const s3 = new Set([3, 4]);

    fn(s1);
    fn(s1);
    fn(s3);

    expect(fn).sinonToBeCalledTimesWith(2, s2);
    expect(fn).not.sinonToBeCalledTimesWith(2, s3);

    expect(() => expect(fn).not.sinonToBeCalledTimesWith(2, s2)).toThrowErrorMatchingSnapshot();
    expect(() => expect(fn).sinonToBeCalledTimesWith(2, s3)).toThrowErrorMatchingSnapshot();
  });

  test('works with Immutable.js objects', () => {
    const fn = sinon.spy();
    const directlyCreated = Immutable.Map([['a', { b: 'c' }]]);
    const indirectlyCreated = Immutable.Map().set('a', { b: 'c' });
    fn(directlyCreated, indirectlyCreated);

    expect(fn).sinonToBeCalledTimesWith(1, indirectlyCreated, directlyCreated);

    expect(() => expect(fn).not.sinonToBeCalledTimesWith(1, indirectlyCreated, directlyCreated)).toThrowErrorMatchingSnapshot();
  });

  test('works with many arguments', () => {
    const fn = sinon.spy();
    fn('foo1', 'bar');
    fn('foo', 'bar');
    fn('foo', 'bar1');
    fn('foo', 'bar');

    expect(fn).sinonToBeCalledTimesWith(2, 'foo', 'bar');

    expect(() => expect(fn).not.sinonToBeCalledTimesWith(2, 'foo', 'bar')).toThrowErrorMatchingSnapshot();
  });

  test("works with many arguments that don't match", () => {
    const fn = sinon.spy();
    fn('foo', 'bar1');
    fn('foo', 'bar');
    fn('foo', 'bar2');
    fn('foo', 'bar3');

    expect(fn).not.sinonToBeCalledTimesWith(2, 'foo', 'bar');

    expect(() => expect(fn).sinonToBeCalledTimesWith(2, 'foo', 'bar')).toThrowErrorMatchingSnapshot();
  });

  test('includes the custom mock name in the error message', () => {
    const fn = sinon.spy();
    fn.named('named-mock');
    fn('foo', 'bar');

    expect(fn).sinonToBeCalledTimesWith(1, 'foo', 'bar');

    expect(() => expect(fn).not.sinonToBeCalledTimesWith(1, 'foo', 'bar')).toThrowErrorMatchingSnapshot();
  });
});
