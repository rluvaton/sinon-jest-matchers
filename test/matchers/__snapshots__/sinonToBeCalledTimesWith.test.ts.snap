// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sinonToBeCalledTimesWith includes the custom mock name in the error message 1`] = `
"expect(named-mock).not.sinonToBeCalledTimesWith(...expected)

Expected not to be called 1 times with: \\"foo\\", \\"bar\\"

Number of calls: 1"
`;

exports[`sinonToBeCalledTimesWith works only on spies or sinon.spy 1`] = `
"expect(received).sinonToBeCalledTimesWith(...expected)

Matcher error: received value must be a stub or spy function

Received has type:  function
Received has value: [Function fn]"
`;

exports[`sinonToBeCalledTimesWith works when not called 1`] = `
"expect(sinon.spy()).sinonToBeCalledTimesWith(...expected)

Expected: \\"foo\\", \\"bar\\"

Expected number of matching calls: 2
Received number of matching calls: 0"
`;

exports[`sinonToBeCalledTimesWith works with Immutable.js objects 1`] = `
"expect(sinon.spy()).not.sinonToBeCalledTimesWith(...expected)

Expected not to be called 1 times with: Immutable.Map {\\"a\\": {\\"b\\": \\"c\\"}}, Immutable.Map {\\"a\\": {\\"b\\": \\"c\\"}}

Number of calls: 1"
`;

exports[`sinonToBeCalledTimesWith works with Map 1`] = `
"expect(sinon.spy()).not.sinonToBeCalledTimesWith(...expected)

Expected not to be called 2 times with: Map {1 => 2, 2 => 1}
Received
       1:     Map {1 => 2, 2 => 1}
       2:     Map {1 => 2, 2 => 1}

Number of calls: 2"
`;

exports[`sinonToBeCalledTimesWith works with Map 2`] = `
"expect(sinon.spy()).sinonToBeCalledTimesWith(...expected)

Expected: Map {\\"a\\" => \\"b\\", \\"b\\" => \\"a\\"}
Received
       1
          Map {
        -   \\"a\\" => \\"b\\",
        -   \\"b\\" => \\"a\\",
        +   1 => 2,
        +   2 => 1,
          },
       2
          Map {
        -   \\"a\\" => \\"b\\",
        -   \\"b\\" => \\"a\\",
        +   1 => 2,
        +   2 => 1,
          },
       3
          Map {\\"a\\" => \\"b\\", \\"b\\" => \\"a\\"},

Expected number of matching calls: 2
Received number of matching calls: 1"
`;

exports[`sinonToBeCalledTimesWith works with Set 1`] = `
"expect(sinon.spy()).not.sinonToBeCalledTimesWith(...expected)

Expected not to be called 2 times with: Set {1, 2}
Received
       1:     Set {1, 2}
       2:     Set {1, 2}

Number of calls: 2"
`;

exports[`sinonToBeCalledTimesWith works with Set 2`] = `
"expect(sinon.spy()).sinonToBeCalledTimesWith(...expected)

Expected: Set {3, 4}
Received
       1
          Set {
        -   3,
        -   4,
        +   1,
        +   2,
          },
       2
          Set {
        -   3,
        -   4,
        +   1,
        +   2,
          },
       3
          Set {3, 4},

Expected number of matching calls: 2
Received number of matching calls: 1"
`;

exports[`sinonToBeCalledTimesWith works with arguments that don't match 1`] = `
"expect(sinon.spy()).sinonToBeCalledTimesWith(...expected)

Expected: \\"foo\\", \\"bar\\"
Received
       1: \\"foo\\", \\"bar\\"
       2: \\"foo\\", \\"bar1\\"

Expected number of matching calls: 2
Received number of matching calls: 1"
`;

exports[`sinonToBeCalledTimesWith works with arguments that don't match in number of arguments 1`] = `
"expect(sinon.spy()).sinonToBeCalledTimesWith(...expected)

Expected: \\"foo\\", \\"bar\\"
Received
       1: \\"foo\\", \\"bar\\"
       2: \\"foo\\", \\"bar\\", \\"plop\\"
       3: \\"foo\\", \\"bar\\"

Expected number of matching calls: 3
Received number of matching calls: 2"
`;

exports[`sinonToBeCalledTimesWith works with arguments that don't match with matchers 1`] = `
"expect(sinon.spy()).sinonToBeCalledTimesWith(...expected)

Expected: Any<String>, Any<Number>
Received
       1: \\"foo\\", 1
       2: \\"foo\\", \\"bar\\"

Expected number of matching calls: 2
Received number of matching calls: 1"
`;

exports[`sinonToBeCalledTimesWith works with arguments that don't match with matchers even when argument is undefined 1`] = `
"expect(sinon.spy()).sinonToBeCalledTimesWith(...expected)

Expected: \\"foo\\", Any<String>
Received: \\"foo\\", undefined

Expected number of matching calls: 1
Received number of matching calls: 0"
`;

exports[`sinonToBeCalledTimesWith works with arguments that match 1`] = `
"expect(sinon.spy()).not.sinonToBeCalledTimesWith(...expected)

Expected not to be called 2 times with: \\"foo\\", \\"bar\\"
Received
       1:     \\"foo\\", \\"bar\\"
       2:     \\"foo\\", \\"bar\\"

Number of calls: 2"
`;

exports[`sinonToBeCalledTimesWith works with arguments that match with matchers 1`] = `
"expect(sinon.spy()).not.sinonToBeCalledTimesWith(...expected)

Expected not to be called 2 times with: Any<String>, Any<String>
Received
       1:     \\"foo\\", \\"bar\\"
       2:     \\"foo\\", \\"bar\\"

Number of calls: 2"
`;

exports[`sinonToBeCalledTimesWith works with many arguments 1`] = `
"expect(sinon.spy()).not.sinonToBeCalledTimesWith(...expected)

Expected not to be called 2 times with: \\"foo\\", \\"bar\\"
Received
       2:     \\"foo\\", \\"bar\\"
       4:     \\"foo\\", \\"bar\\"

Number of calls: 2"
`;

exports[`sinonToBeCalledTimesWith works with many arguments that don't match 1`] = `
"expect(sinon.spy()).sinonToBeCalledTimesWith(...expected)

Expected: \\"foo\\", \\"bar\\"
Received
       1: \\"foo\\", \\"bar1\\"
       2: \\"foo\\", \\"bar\\"
       3: \\"foo\\", \\"bar2\\"

Expected number of matching calls: 2
Received number of matching calls: 1"
`;

exports[`sinonToBeCalledTimesWith works with trailing undefined arguments if requested by the match query 1`] = `
"expect(sinon.spy()).not.sinonToBeCalledTimesWith(...expected)

Expected not to be called 2 times with: \\"foo\\", undefined
Received
       1:     \\"foo\\", undefined
       2:     \\"foo\\", undefined

Number of calls: 2"
`;

exports[`sinonToBeCalledTimesWith works with trailing undefined arguments when explicitly requested as optional by matcher 1`] = `
"expect(sinon.spy()).not.sinonToBeCalledTimesWith(...expected)

Expected not to be called 2 times with: \\"foo\\", optionalFn<>
Received
       1:     \\"foo\\", undefined
       2:     \\"foo\\", undefined

Number of calls: 2"
`;
